pkg/scheduler/core/generic_scheduler.go

```diff
@@ -214,6 +214,22 @@ func (g *genericScheduler) Schedule(pod *v1.Pod, nodeLister algorithm.NodeLister
 		return result, err
 	}
 
+	failedFilterMap := framework.FailedFilterMap{}
+	for nodeName, reasons := range failedPredicateMap {
+		statuses := make([]*framework.Status, len(reasons))
+		for _, reason := range reasons {
+			statuses = append(statuses, framework.NewStatus(framework.Error, reason.GetReason()))
```

Comment:
Not all failed predicates/filters are `Error`. In fact, most of them are `Unschedulable` and we need to distinguish between them.
Also, I don't think it makes sense for us to keep `failedPredicateMap` and pay twice for memory allocation of `failedPredicateMap` and `failedFilterMap` and then pay the cost of conversion to `failedFilterMap`. Instead, we should change `findNodesThatFit` to return a `failedFilterMap`. That way we can populate proper `status.Code` as well.

Reply:
> Not all failed predicates/filters are `Error`. In fact, most of them are `Unschedulable` and we need to distinguish between them.

Does it mean we need a dictionary to translate `PredicateFailureError` to `Status`?

```go
	// ErrDiskConflict is used for NoDiskConflict predicate error.
	ErrDiskConflict = newPredicateFailureError("NoDiskConflict", "node(s) had no available disk")
	// ErrVolumeZoneConflict is used for NoVolumeZoneConflict predicate error.
	ErrVolumeZoneConflict = newPredicateFailureError("NoVolumeZoneConflict", "node(s) had no available volume zone")
	// ErrNodeSelectorNotMatch is used for MatchNodeSelector predicate error.
	ErrNodeSelectorNotMatch = newPredicateFailureError("MatchNodeSelector", "node(s) didn't match node selector")
	// ErrPodAffinityNotMatch is used for MatchInterPodAffinity predicate error.
	ErrPodAffinityNotMatch = newPredicateFailureError("MatchInterPodAffinity", "node(s) didn't match pod affinity/anti-affinity")
	// ErrPodAffinityRulesNotMatch is used for PodAffinityRulesNotMatch predicate error.
	ErrPodAffinityRulesNotMatch = newPredicateFailureError("PodAffinityRulesNotMatch", "node(s) didn't match pod affinity rules")
	// ErrPodAntiAffinityRulesNotMatch is used for PodAntiAffinityRulesNotMatch predicate error.
	ErrPodAntiAffinityRulesNotMatch = newPredicateFailureError("PodAntiAffinityRulesNotMatch", "node(s) didn't match pod anti-affinity rules")
....
```

> Also, I don't think it makes sense for us to keep `failedPredicateMap` and pay twice for memory allocation of `failedPredicateMap` and `failedFilterMap` and then pay the cost of conversion to `failedFilterMap`. Instead, we should change `findNodesThatFit` to return a `failedFilterMap`. That way we can populate proper `status.Code` as well.

The cost is inevitable since the predicate function returns `PredicateFailureReason` in the first place. We can't make it disappear by returning `failedFilterMap` in `findNodesThatFit`, the cost just moves from `Schedule` to `findNodesThatFit` or `podFitsOnNode`.

Reply:
I don't agree with this dependency, converting FailurePredicate to Status looks like a hack to me that trickled to some strange logic that is hard to reason about across the whole file. I think we should return two separate maps, one for the predicates and one for the filters. As we move predicates to filters, the former will get naturally deprecated. The cost of allocating a second map shouldn't be a concern since we are not really duplicating any state I think.

Reply:
> I don't agree with this dependency

Could you please elaborate which dependency you are referring to? We will have dependency on "framework" here anyway.

Reply:
Thanks Bobby, I guess what I meant is the relation we are creating between FailurePredicate and framework.Status. I agree we will have a dependency on the framework in all cases.

Reply:
@ahg-g I see your point about returning a predicate name in the `Message` field of status and using what is supposed be an arbitrary string as a key in a map is not clean. That's a fair point.

This also made me realize that we must let `Filter` plugins tell us whether they can be resolved by preemption or not.

I think we can stick to the existing interface of returning `FailedPredicateMap` while we are transitioning to `Filter` plugins. Sorry @draveness for this change of opinion!